x/plans/keeper/migrations.go

```diff
@@ -22,3 +25,39 @@ func (m Migrator) Migrate2to3(ctx sdk.Context) error {
 	}
 	return nil
 }
+
+func (m Migrator) Migrate3to4(ctx sdk.Context) error {
+	planIndices := m.keeper.GetAllPlanIndices(ctx)
+
+	for _, planIndex := range planIndices {
+		var oldPlanStruct v3.Plan
+		if found := m.keeper.plansFS.FindEntry(ctx, planIndex, uint64(ctx.BlockHeight()), &oldPlanStruct); !found {
```

Comment:
With the list of blocks from the previous comment, you can call plansFS.ReadEntry() to get each specific version (it avoids the iterator lookup).